src/nvim/lua/base64.c

```diff
@@ -0,0 +1,257 @@
+  uint32_t ret = 0;
+  for (size_t i = 4; i; i--) {
+    ret |= ((uint32_t)buf[i - 1]) << ((4 - i) * 8);
+  }
+  return ret;
+# endif
+}
+#endif
+
+static char *base64_encode(const char *src, size_t src_len)
+{
+  assert(src != NULL);
+
+  const size_t out_len = ((src_len + 2) / 3) * 4;
+  char *dest = xmalloc(out_len + 1);
+
+  size_t src_i = 0;
+  size_t out_i = 0;
+
+  // Read 8 bytes at a time as much as we can
+  for (; src_i + 7 < src_len; src_i += 6) {
+    uint64_t bits_h;
+    memcpy(&bits_h, &src[src_i], sizeof(uint64_t));
+    const uint64_t bits_be = htobe64(bits_h);
+    dest[out_i + 0] = alphabet[(bits_be >> 58) & 0x3F];
+    dest[out_i + 1] = alphabet[(bits_be >> 52) & 0x3F];
+    dest[out_i + 2] = alphabet[(bits_be >> 46) & 0x3F];
+    dest[out_i + 3] = alphabet[(bits_be >> 40) & 0x3F];
+    dest[out_i + 4] = alphabet[(bits_be >> 34) & 0x3F];
+    dest[out_i + 5] = alphabet[(bits_be >> 28) & 0x3F];
+    dest[out_i + 6] = alphabet[(bits_be >> 22) & 0x3F];
+    dest[out_i + 7] = alphabet[(bits_be >> 16) & 0x3F];
+    out_i += sizeof(uint64_t);
+  }
+
+  for (; src_i + 3 < src_len; src_i += 3) {
+    uint32_t bits_h;
+    memcpy(&bits_h, &src[src_i], sizeof(uint32_t));
+    const uint32_t bits_be = htobe32(bits_h);
+    dest[out_i + 0] = alphabet[(bits_be >> 26) & 0x3F];
+    dest[out_i + 1] = alphabet[(bits_be >> 20) & 0x3F];
+    dest[out_i + 2] = alphabet[(bits_be >> 14) & 0x3F];
+    dest[out_i + 3] = alphabet[(bits_be >> 8) & 0x3F];
+    out_i += sizeof(uint32_t);
+  }
+
+  if (src_i + 2 < src_len) {
+    dest[out_i + 0] = alphabet[src[src_i] >> 2];
+    dest[out_i + 1] = alphabet[((src[src_i] & 0x3) << 4) | (src[src_i + 1] >> 4)];
+    dest[out_i + 2] = alphabet[(src[src_i + 1] & 0xF) << 2 | (src[src_i + 2] >> 6)];
+    dest[out_i + 3] = alphabet[(src[src_i + 2] & 0x3F)];
+    out_i += 4;
+  } else if (src_i + 1 < src_len) {
+    dest[out_i + 0] = alphabet[src[src_i] >> 2];
+    dest[out_i + 1] = alphabet[((src[src_i] & 0x3) << 4) | (src[src_i + 1] >> 4)];
+    dest[out_i + 2] = alphabet[(src[src_i + 1] & 0xF) << 2];
+    out_i += 3;
+  } else if (src_i < src_len) {
+    dest[out_i + 0] = alphabet[src[src_i] >> 2];
+    dest[out_i + 1] = alphabet[(src[src_i] & 0x3) << 4];
+    out_i += 2;
+  }
+
+  for (; out_i < out_len; out_i++) {
+    dest[out_i] = '=';
+  }
+
+  dest[out_len] = '\0';
+
+  return dest;
+}
+
+static char *base64_decode(const char *src, size_t src_len)
+{
+  assert(src != NULL);
+
+  char *dest = NULL;
+
+  if (src_len % 4 != 0) {
+    goto invalid;
+  }
+
+  size_t out_len = (src_len / 4) * 3;
+  if (src_len >= 1 && src[src_len - 1] == '=') {
+    out_len--;
+  }
+  if (src_len >= 2 && src[src_len - 2] == '=') {
+    out_len--;
+  }
+
+  dest = xmalloc(out_len + 1);
+
+  int acc = 0;
+  int acc_len = 0;
+  size_t out_i = 0;
+  size_t src_i = 0;
+  int leftover_i = -1;
+
+  for (; src_i < src_len; src_i++) {
+    const char c = src[src_i];
+    const uint8_t d = char_to_index[c & 0xFF];
```

Comment:
`c` is already `char`
```suggestion
    const uint8_t d = char_to_index[(uint8_t)c];
```

Reply:
452da56afa706a8da2f0a2948014c8fb6ebb515d